arxivdigest/frontend/database/general.py

```diff
@@ -260,3 +267,64 @@ def insertFeedback(user_id, article_id, type, feedback_text):
             return "Unknown article id."
         raise
     conn.commit()
+
+
+def get_keyword_feedback_user(user_id):
+    """ Gets all the feedback on keywords from given user.
+    :param user_id: User to get feedback for.
+    :return: List of feedback instances.
+    """
+    cur = getDb().cursor(dictionary=True)
+    sql = '''SELECT keyword, feedback, datestamp FROM  keyword_feedback 
+             WHERE user_ID = %s'''
+    cur.execute(sql, (user_id,))
+    return cur.fetchall()
+
+
+def get_feedback(user_id):
+    """Get feedback from given user.
+    :param user_id: User to get feedback for.
+    :return: List of feedback instances.
+    """
+    cur = getDb().cursor(dictionary=True)
+    sql = '''SELECT article_ID, type, feedback_text 
+             FROM feedback WHERE user_ID = %s'''
+    cur.execute(sql, (user_id,))
+    return cur.fetchall()
+
+
+def get_system_recommendations(user_id):
+    """Get system recommendations for given user. Includes user interactiondata
+    if the
+    :param user_id: User to get feedback for.
+    :return: List of system recommendation instances.
+    """
+    cur = getDb().cursor(dictionary=True)
+    sql = '''SELECT sr.article_ID, s.system_name, sr.explanation, 
+             sr.score AS system_score, ur.score AS recommendation_order,
+             ur.seen_email, ur.seen_web, ur.clicked_email, ur.clicked_web,
+             ur.liked, sr.recommendation_date
+             FROM system_recommendations sr 
+             NATURAL JOIN systems s
+             LEFT JOIN user_recommendations ur 
+             ON sr.article_ID = ur.article_ID 
+             AND sr.user_ID = ur.user_ID
+             AND sr.system_ID = ur.system_ID
+             WHERE sr.user_ID = %s
+             ORDER BY sr.recommendation_date desc,
+             s.system_name desc, sr.score desc'''
+    cur.execute(sql, (user_id,))
+    return cur.fetchall()
+
+
+def get_all_userdata(user_id):
+    """Get all data for the given user as a dictionary.
+
+    :param user_id: Id of the user to retrieve data for.
+    :return: Dictionary of userdata.
+    """
+
+    return {'user': get_user(user_id),
+            'keyword_feedback': get_keyword_feedback_user(user_id),
```

Comment:
Can we group all data under `topic_recommendations` and `article_recommendations`?
If there was free text feedback, that should be associated with the respective article recommendation.

Reply:
When we change how topics work this information will probably be included under `topic_recommendations`, but at the moment there is nothing to combine it with.

The other feedback field contains feedback that can be linked to articles or topics or neither so we cannot remove it completely. Some of the feedback could maybe be moved under their respective keywords or articles, but I'm not sure if this would be more organized.

Reply:
From what I understand, `get_feedback()` returns free text feedback associated with article recommendations, so there is always an article_id

Reply:
At the moment we have 5 different freetext feedback types that uses this table:

 *  Explanation - connected to article
 *  Recommendation - connected to article
 *  Bug - Report bug, not connected to article
 *  Request Feature, not connected to article
 *  Other, not connected to article

In the future we'll also add topic feedback here, this will also not be connected to an article.

We could sort the ones connected to a article under their relevant article and the same for topics.
But there would still be the other types of feedback left.

Reply:
I understand, OK. Let's keep the original structure then, but call the three fields `article_recommendations`, `topic_recommendations`, and `freetext_feedback`. Non-free text feedback (clicks & likes for articles and approve/reject for topic recommendations) should go under article/topic_recommendations.

Reply:
Renamed fields as specified, although topic_recommendation won't be populated correctly until the upcoming topic rework. Also added systems as a field now that systems are tied to a user.